logic/scorer/scorer.go

```diff
@@ -60,260 +34,247 @@ func points(lead model.Card, hand []model.Card, isCrib bool) int {
-		if values[1] == values[3] {
-			// we know that values[0] = values[1] - 1 and values[3] = values[4] + 1
-			// because we already checked quads
-			// A|B B B|C
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		} else if values[2] == min {
-			// A A A|B|C
-			// A A A|C C
-			if values[3] == max {
-				// triple/pair combo
-				return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-			}
-			// triple run of 3 is worth 15
-			// because 3 runs of three (worth 9) and a triple for 6
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		} else if values[2] == max {
-			// A|B|C C C
-			// A A|C C C
-			if values[1] == min {
-				return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-			}
-			return tripleRunOfThree, 15 /* triple run of 3 */
-		}
-
-		// If the hand doesn't have a quad or a triple, then it must be a double-double run of three
-		return doubleDoubleRunOfThree, 16 /* double double run of 3 */
+func (fs *flushScorer) atEach(c model.Card) {
+	if c == fs.leadCard {
+		return
+	}
+	if c.Suit != fs.firstCard.Suit {
+		fs.isHandFlush = false
 	}
+}
 
-	// checking triplets for remaining hands are all the same
-	// look for (triple/pair, triple/odds)
-	if values[1] == values[3] {
-		// this is definitely a triple without a pair
-		// A|B B B|D
-		return triplet, 6 /* triplet */
-	} else if values[2] == min {
-		// A A A|B|D
-		// A A A|D D
-		if values[3] == max {
-			// triple/pair combo
-			return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-		}
-		return triplet, 6 /* triplet */
-	} else if values[2] == max {
-		// A|B|D D D
-		// A A|D D D
-		if values[1] == min {
-			return triplet | onepair, 8 /* 6 for a triple, 2 for a pair */
-		}
-		return triplet, 6 /* triplet */
+func (fs *flushScorer) accumulate() (scoreType, uint8) {
+	if !fs.isHandFlush {
+		return 0, 0
+	}
+	if fs.firstCard.Suit == fs.leadCard.Suit {
+		return flush5, 5
+	}
+	if fs.isCrib {
+		return 0, 0
 	}
+	return flush4, 4
+}
 
-	switch diffOf5Cards {
-	case 3:
-		// If we have a diff of 3, then we either have a two pair or a double run of 4
-		// check double run of four: A A|B|C|D
-		// check two-pair          : A A|B B|D
-		numPairs := 0
-		for i := 1; i < len(values); i++ {
-			if values[i] == values[i-1] {
-				numPairs++
-			}
-		}
+type iterateHandResult struct {
+	valuesToCounts [valuesToCountsLength]uint8
+	values         [numCardsToScore]int
+	ptValues       [numCardsToScore]int
+	scoreType      scoreType
+	totalPoints    uint8
+}
 
-		if numPairs == 2 {
-			// we found two pairs, which mean it cannot be a double run of 4
-			return twopair, 4 /* two pair */
+// iterateHand does all the scoring possible with a single iteration through the hand
+// we can't do fifteens or runs here directly, but the information returned will aid in doing those efficiently
+func iterateHand(lead model.Card, hand []model.Card, isCrib bool) iterateHandResult {
+	res := iterateHandResult{}
+	hasNobs := false
+	ps := &pairScorer{}
```

Comment:
:mag: how does this _not_ escape to heap?

Reply:
exercise left up to the reader

Reply:
the answer is that it's fully scoped so that the go compiler knows that it doesn't _need_ to escape. but that was a test for you;)